cleverhans/model.py

```diff
@@ -0,0 +1,105 @@
+        # The values output_dict are also a dictionary mapping layer names
+        # to symbolic representation of the output of that layer.
+        self.fprop_cache = {}
+
+        # By default, we assume the model is being used for training (i.e.,
+        # 'train' time). If the model is used for inference or in a different
+        # state, a call to set_state() should be made first.
+        self._state = 'train'
+
+        pass
+
+    @property
+    def state(self):
+        return self._state
+
+    @state.setter
+    def state(self, state_val):
+        self._state = state_val
+
+    def __call__(self, *args, **kwargs):
+        """
+        For compatibility with functions used as model definitions (taking
+        an input tensor and returning the tensor giving the output
+        of the model on that input).
+        """
+        return self.get_probs(*args, **kwargs)
+
+    def get_layer(self, x, layer):
+        """
+        Expose the hidden features of a model given a layer name.
+        :param x: A symbolic representation of the network input
+        :param layer: The name of the hidden layer to return features at.
+        :return: A symbolic representation of the hidden features
+        """
+        # Return the symbolic representation for this layer.
+        return self.fprop(x)[layer]
+
+    def get_logits(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output logits (i.e., the
+                 values fed as inputs to the softmax layer).
+        """
+        return self.get_layer(x, 'logits')
+
+    def get_probs(self, x):
+        """
+        :param x: A symbolic representation of the network input
+        :return: A symbolic representation of the output probabilities (i.e.,
+                the output values produced by the softmax layer).
+        """
+        if 'probs' in self.get_layer_names():
+            return self.get_layer(x, 'probs')
+        else:
+            import tensorflow as tf
+            return tf.nn.softmax(self.get_logits(x))
+
+    def get_layer_names(self):
+        """
+        :return: a list of names for the layers that can be exposed by this
+        model abstraction.
+        """
+        raise NotImplementedError('`get_layer_names` not implemented.')
+
+    def fprop(self, x):
+        """
+        Exposes all the layers of the model returned by get_layer_names.
+        :param x: A symbolic representation of the network input
+        :return: A dictionary mapping layer names to the symbolic
+                 representation of their output.
+        """
+        # In case of cache hit, return cached dictionary of output tensors.
+        if (x, self.state) in self.fprop_cache.keys():
+            return self.fprop_cache[(x, self.state)]
+        else:
+            result = self._fprop(x)
+            self.fprop_cache[(x, self.state)] = result
```

Comment:
I'm confused about why we need the cache now. I thought the point of the cache was to avoid rebuilding TensorFlow graphs when people made more calls to generate_np. But this is for symbolic calls, not numpy calls. Is there a use case where we expect people to repeatedly request the same graph explicitly?